File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/views/RawOutputConsole.java
package isabelle.eclipse.ui.views;

import java.io.IOException;
import isabelle.Isabelle_Process.Output;
import isabelle.Event_Bus;
import isabelle.Session;
import isabelle.XML;
import isabelle.eclipse.core.util.SafeSessionActor;
import isabelle.eclipse.core.util.SessionEventSupport;
import isabelle.eclipse.ui.IsabelleUIPlugin;
import isabelle.scala.ISessionRawMessageListener;
import isabelle.scala.ScalaCollections;
import org.eclipse.jface.resource.ImageDescriptor;
import org.eclipse.ui.console.IOConsoleOutputStream;
import org.eclipse.ui.console.MessageConsole;

import scala.actors.Actor;


public class RawOutputConsole extends MessageConsole {
	
	private final SessionEventSupport<?> sessionEvents;

	private IOConsoleOutputStream consoleStream;
	
	public RawOutputConsole(String name, ImageDescriptor imageDescriptor) {
		super(name, imageDescriptor, true);
		
		sessionEvents = new SessionEventSupport<Output>() {
			
			@Override
			public Actor sessionActor() {
				return (Actor) new SafeSessionActor().rawMessages(new ISessionRawMessageListener() {
					
					@Override
					public void handleMessage(Output result) {
						outputMessage(result);
					}
				}).getActor();
			}
			
			@Override
			public scala.collection.immutable.List<Event_Bus<Output>> sessionEvents0(Session session) {
				return ScalaCollections.singletonList(session.raw_output_messages());
			}
		};
		sessionEvents.init();
	}

	@Override
	protected void init() {
		super.init();

		if (consoleStream == null) {
			consoleStream = this.newOutputStream();
		}
		
		try {
			consoleStream.write("Starting Raw Output Console");
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	@Override
	protected void dispose() {
		
		sessionEvents.dispose();
		
		if (consoleStream != null) {
			try {
				consoleStream.close();
			} catch (IOException e) {
				IsabelleUIPlugin.log("Unable to close raw output console", e);
			} finally {
				consoleStream = null;
			}
			
		}
		
		super.dispose();
	}
	
	private void outputMessage(Output result) {

		if (consoleStream == null) {
			return;
		}
		
		if (!result.is_stdout()) {
			return;
		}
		
		String output = XML.content(result.message()).mkString();
		
		try {
			consoleStream.write(output);
		} catch (IOException e) {
			IsabelleUIPlugin.log("Problems writing to raw output console", e);
		}
		
	}


}

File: isabelle.eclipse.ui/src/isabelle/eclipse/ui/views/RawOutputConsole.scala
package isabelle.eclipse.ui.views

import scala.actors.Actor.loop
import scala.actors.Actor.react
import scala.util.Try

import org.eclipse.jface.resource.ImageDescriptor
import org.eclipse.ui.console.IConsole
import org.eclipse.ui.console.MessageConsole

import isabelle.Isabelle_Process
import isabelle.Session
import isabelle.XML
import isabelle.eclipse.core.util.LoggingActor
import isabelle.eclipse.core.util.SessionEvents
import isabelle.eclipse.ui.internal.IsabelleImages
import isabelle.eclipse.ui.internal.IsabelleUIPlugin
import isabelle.eclipse.ui.internal.IsabelleUIPlugin.error
import isabelle.eclipse.ui.internal.IsabelleUIPlugin.log


/**
 * A message console that tracks Isabelle Raw Output messages when initialised.
 * 
 * @author Andrius Velykis
 */
class RawOutputConsole(name: String, consoleType: String, image: ImageDescriptor) 
    extends MessageConsole(name, consoleType, image, true) with SessionEvents {

  // the actor to react to session events
  override protected val sessionActor = LoggingActor {
    loop {
      react {
        case output: Isabelle_Process.Output =>
          consoleStream.print(XML.content(output.message))
          if (!output.is_stdout && !output.is_stderr) consoleStream.println()

        case bad => System.err.println("RawOutputConsole: ignoring bad message " + bad)
      }
    }
  }

  // subscribe to commands change session events
  override protected def sessionEvents(session: Session) = List(session.raw_output_messages)
  
  private lazy val consoleStream = newMessageStream()

  override protected def init() {
    super.init()

    consoleStream.println("Starting Raw Output Console")
    initSessionEvents()
  }


  override protected def dispose() {
    disposeSessionEvents()

    Try(consoleStream.close()).failed foreach ( ex =>
      log(error(Some(ex), Some("Unable to close raw output console"))) )

    super.dispose()
  }

}

class RawOutputConsoleFactory extends SingletonConsoleFactory {

  val consoleType = IsabelleUIPlugin.plugin.pluginId + ".rawOutputConsole"

  override def createConsole(): IConsole =
    new RawOutputConsole("Isabelle Raw Output", consoleType, IsabelleImages.RAW_OUTPUT_CONSOLE)

}
